kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs

```diff
@@ -19,44 +20,77 @@ use tracing::{debug, warn};
 
 impl<'tcx> GotocCtx<'tcx> {
     fn codegen_comparison(&mut self, op: &BinOp, e1: &Operand<'tcx>, e2: &Operand<'tcx>) -> Expr {
-        match op {
-            BinOp::Eq => {
-                if self.operand_ty(e1).is_floating_point() {
-                    self.codegen_operand(e1).feq(self.codegen_operand(e2))
-                } else {
-                    self.codegen_operand(e1).eq(self.codegen_operand(e2))
-                }
-            }
-            BinOp::Lt => self.codegen_operand(e1).lt(self.codegen_operand(e2)),
-            BinOp::Le => self.codegen_operand(e1).le(self.codegen_operand(e2)),
-            BinOp::Ne => {
-                if self.operand_ty(e1).is_floating_point() {
-                    self.codegen_operand(e1).fneq(self.codegen_operand(e2))
-                } else {
-                    self.codegen_operand(e1).neq(self.codegen_operand(e2))
-                }
-            }
-            BinOp::Ge => self.codegen_operand(e1).ge(self.codegen_operand(e2)),
-            BinOp::Gt => self.codegen_operand(e1).gt(self.codegen_operand(e2)),
-            _ => unreachable!(),
-        }
+        let left = self.codegen_operand(e1);
+        let right = self.codegen_operand(e2);
+        let is_float = self.operand_ty(e1).is_floating_point();
+        comparison_expr(op, left, right, is_float)
     }
 
     /// This function codegen comparison for fat pointers.
-    /// We currently don't support this. See issue #327 for more information.
+    /// Fat pointer comparison must compare the raw data pointer as well as its metadata portion.
+    ///
+    /// Since vtable pointer comparison is not well defined and it has many nuances, we decided to
+    /// fail if the user code performs such comparison.
+    ///
+    /// See https://github.com/model-checking/kani/issues/327 for more details.
     fn codegen_comparison_fat_ptr(
         &mut self,
         op: &BinOp,
         left: &Operand<'tcx>,
         right: &Operand<'tcx>,
     ) -> Expr {
         debug!(?op, ?left, ?right, "codegen_comparison_fat_ptr");
-        self.codegen_unimplemented(
-            &format!("Fat pointer comparison '{:?}'", op),
-            Type::Bool,
-            Location::None,
-            "https://github.com/model-checking/kani/issues/327",
-        )
+        let left_typ = self.operand_ty(left);
+        let right_typ = self.operand_ty(left);
+        assert_eq!(left_typ, right_typ, "Cannot compare pointers of different types");
+        assert!(self.is_ref_of_unsized(left_typ));
+
+        if self.is_vtable_fat_pointer(left_typ) {
+            // Codegen an assertion failure since vtable comparison is not stable.
+            let ret_type = Type::Bool;
+            let body = vec![
+                self.codegen_assert_false(
+                    PropertyClass::KaniCheck,
+                    format!("Reached unstable vtable comparison '{:?}'", op).as_str(),
+                    Location::None,
```

Comment:
Is there a way to get the locations for these statements? If so, I suggest adding them.

Reply:
I can try to refactor things a bit to pass down the location.